relayer/relayer/src/handlers/eth.rs

```diff
@@ -0,0 +1,206 @@
+use log::{debug, error, info, trace, warn};
+use thiserror::Error;
+use tokio::{
+    select,
+    sync::{broadcast, mpsc},
+    time::sleep,
+};
+
+use crate::{
+    config::Config,
+    connections::azero::AzeroConnectionWithSigner,
+    contracts::{AzeroContractError, CrosschainTransferRequestFilter, MostEvents, MostInstance},
+    helpers::concat_u8_arrays,
+    listeners::{EthMostEvents, ALEPH_BLOCK_PROD_TIME_SEC},
+    CircuitBreakerEvent,
+};
+
+#[derive(Debug, Error)]
+#[error(transparent)]
+#[non_exhaustive]
+pub enum EthereumEventHandlerError {
+    #[error("azero contract error")]
+    AzeroContract(#[from] AzeroContractError),
+
+    #[error("receive_request tx has failed")]
+    ReceiveRequestTxFailure {
+        request_hash: String,
+        committee_id: u128,
+        dest_token_address: String,
+        amount: u128,
+        dest_receiver_address: String,
+        request_nonce: u128,
+    },
+}
+
+pub struct EthereumEventHandler;
+
+impl EthereumEventHandler {
+    pub async fn handle_event(
+        event: MostEvents,
+        config: &Config,
+        azero_connection: &AzeroConnectionWithSigner,
+    ) -> Result<(), EthereumEventHandlerError> {
+        let Config {
+            azero_contract_address,
+            azero_contract_metadata,
+            ..
+        } = config;
+
+        if let MostEvents::CrosschainTransferRequestFilter(
+            crosschain_transfer_event @ CrosschainTransferRequestFilter {
+                committee_id,
+                dest_token_address,
+                amount,
+                dest_receiver_address,
+                request_nonce,
+                ..
+            },
+        ) = event
+        {
+            info!("handling eth contract event: {crosschain_transfer_event:?}");
+
+            // concat bytes
+            let bytes = concat_u8_arrays(vec![
+                &committee_id.as_u128().to_le_bytes(),
+                &dest_token_address,
+                &amount.as_u128().to_le_bytes(),
+                &dest_receiver_address,
+                &request_nonce.as_u128().to_le_bytes(),
+            ]);
+
+            trace!("event concatenated bytes: {bytes:?}");
+
+            let request_hash = keccak256(bytes);
+            info!("hashed event encoding: {request_hash:?}");
+
+            let contract = MostInstance::new(
+                azero_contract_address,
+                azero_contract_metadata,
+                config.azero_ref_time_limit,
+                config.azero_proof_size_limit,
+            )?;
+
+            // send vote
+
+            let committee_id = committee_id.as_u128();
+            let amount = amount.as_u128();
+            let request_nonce = request_nonce.as_u128();
+
+            while contract
+                .needs_signature(
+                    azero_connection.as_connection(),
+                    request_hash,
+                    azero_connection.account_id().clone(),
+                    committee_id,
+                )
```

Comment:
There is this question of which block is used to check this -- best or finalized? In theory if best is used (which it most likely is, but one would need to check), then this could actually behave incorrectly.